### ISSUE TITLE
Errors detected in syntax trees parsed with esprima

### ISSUE THREAD

**Issue by joshuawarner32** [estools/esvalid#50]
TL;DR: `try { esvalid.isValid(esprima.parse(text)) } catch(e) { true }` should, intuitively, always evaluate to `true`.  This is about some cases where it doesn't.  This is not _necessarily_ a bug report.

I wrote a quick tool to slurp up all the javascript files on my harddrive, run it through esprima and esvalid, select the files that parsed fine but failed the esvalid checks, and automatically minify them.  Oh yeah, and unique them by the particular error I get from esvalid.  Sorry, I _may_ have had a little too much fun with this. ;)

Here's what I found:
- 10 instances of "unrecognized node type", all of which are ES6-specific:  ForOfStatement, ClassDeclaration, TemplateLiteral, TaggedTemplateExpression, Super, YieldExpression, ArrowFunctionExpression, ClassExpression, ObjectPattern, SpreadElement
- 2 instances of ES6 features that aren't explicitly new node types
- 17 instances of fallout from the unrecognized node types (i.e. "cascading failures")
- 4 instances of "identical keys" that the esprima isn't catching, but could
- an instance where esprima doesn't detect a break without a switch, where it could
- **10 instances where `function(){}` is placed where expressions/statements are expected.**

That last case is the most interesting to me.  For example: `for(;;)function f(){}", "switch(n){case 1:function f(){}}", etc.  Here, I think either:
- Esprima is really generating invalid trees
- The examples are really syntax errors, and esprima is being lenient, not catching some syntax errors it could
- Esvalid is wrong, and that particular syntax tree is perfectly valid.

What are your thoughts?

---

You can see the full results here: https://gist.github.com/joshuawarner32/994d2756687551d30d1d

**Comment by michaelficarra** [comment]
> - 10 instances of "unrecognized node type", all of which are ES6-specific

Expected. esvalid has not been updated with support for the ES6 extensions to ESTree.

> - 2 instances of ES6 features that aren't explicitly new node types

Also expected for the same reason as above.

> - 17 instances of fallout from the unrecognized node types

Not much we can do about that if we're trying to continue with best effort.

> - 4 instances of "identical keys" that the esprima isn't catching, but could
> - an instance where esprima doesn't detect a break without a switch, where it could

You should report the issues, but I'd be surprised if they weren't already aware of it.

> - 10 instances where function(){} is placed where expressions/statements are expected.

Yes, this is something that was _not_ allowed according to ES5, but supported (with varying semantics) in most implementations. esprima parses it, knowing it is a non-standard extension to the language. As of ES6, functions are allowed in certain additional positions (inside block statements, labelled statements, IfStatements, switches, try/catch/finally blocks), but still not allowed in IterationStatements.

Thanks for running these test. Looks like the only action item here is to add ES6 support, which can be tracked in a dedicated issue.
